File: scripts/tools/ghapp_download_logs.py
#!/usr/bin/env python3
"""
Download GitHub Actions run logs using a GitHub App.

Requires:
 - App ID (integer)
 - App private key (PEM file path)
 - Repository owner/name
 - Run ID

Examples:
  python scripts/tools/ghapp_download_logs.py \
    --app-id 2032160 \
    --key-file gametp-ai-bot.2025-10-10.private-key.pem \
    --owner skyoxu --repo gametp --run-id 18412530982 --unzip

This script attempts to install its Python dependencies (pyjwt, cryptography, requests)
if they are missing. It prints a short error summary after download.
"""
import argparse
import base64
import json
import os
import subprocess
import sys
import time
from pathlib import Path


def ensure_deps():
    try:
        import jwt  # noqa: F401
        import requests  # noqa: F401
    except Exception:
        # Install quietly; fall back to non-quiet if needed
        pkgs = ["pyjwt", "cryptography", "requests"]
        try:
            subprocess.check_call([sys.executable, "-m", "pip", "install", "-q"] + pkgs)
        except subprocess.CalledProcessError:
            subprocess.check_call([sys.executable, "-m", "pip", "install"] + pkgs)


def b64u(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def make_jwt(app_id: str, pem_path: Path) -> str:
    import jwt  # type: ignore

    private_key = pem_path.read_text(encoding="utf-8")
    now = int(time.time())
    payload = {
        "iat": now - 60,  # allow clock skew
        "exp": now + 9 * 60,  # 9 minutes
        "iss": app_id,
    }
    token = jwt.encode(payload, private_key, algorithm="RS256")
    if isinstance(token, bytes):
        token = token.decode("ascii")
    return token


def github_get(url: str, headers: dict) -> tuple[int, dict, bytes]:
    import requests  # type: ignore

    resp = requests.get(url, headers=headers)
    try:
        js = resp.json() if resp.headers.get("content-type", "").startswith("application/json") else {}
    except Exception:
        js = {}
    return resp.status_code, js, resp.content


def github_post(url: str, headers: dict, data: dict | None = None) -> tuple[int, dict, bytes]:
    import requests  # type: ignore

    resp = requests.post(url, headers=headers, json=data or {})
    try:
        js = resp.json() if resp.headers.get("content-type", "").startswith("application/json") else {}
    except Exception:
        js = {}
    return resp.status_code, js, resp.content


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--app-id", required=True)
    parser.add_argument("--key-file", required=True)
    parser.add_argument("--owner", required=True)
    parser.add_argument("--repo", required=True)
    parser.add_argument("--run-id", required=True)
    parser.add_argument("--out", default=None)
    parser.add_argument("--unzip", action="store_true")
    args = parser.parse_args()

    ensure_deps()

    pem_path = Path(args.key_file)
    if not pem_path.exists():
        print(f"[ERR] Private key not found: {pem_path}", file=sys.stderr)
        return 2

    jwt_token = make_jwt(args.app_id, pem_path)

    # 1) Find installation for the repo
    api = "https://api.github.com"
    repo_install_url = f"{api}/repos/{args.owner}/{args.repo}/installation"
    gh_headers_jwt = {
        "Authorization": f"Bearer {jwt_token}",
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
    }
    sc, js, _ = github_get(repo_install_url, gh_headers_jwt)
    if sc != 200:
        print(f"[ERR] Failed to resolve installation (HTTP {sc}): {js}", file=sys.stderr)
        return 3
    inst_id = js.get("id")
    if not inst_id:
        print("[ERR] No installation id in response", file=sys.stderr)
        return 3

    # 2) Create installation token
    tok_url = f"{api}/app/installations/{inst_id}/access_tokens"
    sc, js, _ = github_post(tok_url, gh_headers_jwt, data={})
    if sc != 201:
        print(f"[ERR] Failed to create installation token (HTTP {sc}): {js}", file=sys.stderr)
        return 4
    inst_token = js.get("token")
    if not inst_token:
        print("[ERR] No token in installation token response", file=sys.stderr)
        return 4

    # 3) Download run logs zip
    logs_url = f"{api}/repos/{args.owner}/{args.repo}/actions/runs/{args.run_id}/logs"
    gh_headers_run = {
        "Authorization": f"Bearer {inst_token}",
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
    }
    import requests  # type: ignore

    resp = requests.get(logs_url, headers=gh_headers_run, stream=True)
    if resp.status_code != 200:
        try:
            js = resp.json()
        except Exception:
            js = {"raw": resp.text[:200]}
        print(f"[ERR] Download logs failed (HTTP {resp.status_code}): {js}", file=sys.stderr)
        return 5

    out = args.out or f"run-{args.run_id}.zip"
    with open(out, "wb") as f:
        for chunk in resp.iter_content(chunk_size=8192):
            if chunk:
                f.write(chunk)
    print(f"[OK] Logs downloaded: {out}")

    if args.unzip:
        try:
            import zipfile
            dest = Path(f"logs-{args.run_id}")
            if dest.exists():
                # Do not delete automatically; write into existing
                pass
            with zipfile.ZipFile(out, "r") as z:
                z.extractall(dest)
            print(f"[OK] Logs extracted to: {dest}")

            # Heuristic error summary
            err_lines = []
            for p in dest.rglob("*.txt"):
                try:
                    for ln in p.read_text(encoding="utf-8", errors="ignore").splitlines():
                        low = ln.lower()
                        if any(k in low for k in (" error ", "error:", "failed", "traceback", "exception")):
                            err_lines.append((p.name, ln.strip()))
                            if len(err_lines) >= 60:
                                break
                    if len(err_lines) >= 60:
                        break
                except Exception:
                    continue
            if err_lines:
                print("\n[Error Summary, top matches]:")
                for name, ln in err_lines[:40]:
                    print(f"  [{name}] {ln}")
            else:
                print("[Info] No obvious error lines found in extracted logs.")
        except Exception as e:
            print(f"[WARN] Unzip/summary failed: {e}")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/tools/ghapp-download-logs.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import https from 'node:https';

function b64u(buf) {
  return Buffer.from(buf)
    .toString('base64')
    .replace(/=/g, '')
    .replace(/\+/g, '-')
    .replace(/\//g, '_');
}

function makeJwt(appId, pem) {
  const header = { alg: 'RS256', typ: 'JWT' };
  const now = Math.floor(Date.now() / 1000);
  const payload = { iat: now - 60, exp: now + 9 * 60, iss: String(appId) };
  const enc = `${b64u(JSON.stringify(header))}.${b64u(JSON.stringify(payload))}`;
  const sign = crypto.createSign('RSA-SHA256');
  sign.update(enc);
  const sig = sign.sign(pem);
  return `${enc}.${b64u(sig)}`;
}

function requestJson(method, url, headers, body) {
  return new Promise((resolve, reject) => {
    const u = new URL(url);
    const opts = {
      method,
      hostname: u.hostname,
      path: u.pathname + u.search,
      headers: { 'User-Agent': 'ghapp-client', ...headers },
    };
    const req = https.request(opts, res => {
      const chunks = [];
      res.on('data', c => chunks.push(c));
      res.on('end', () => {
        const buf = Buffer.concat(chunks);
        const ct = String(res.headers['content-type'] || '');
        let js = {};
        if (ct.startsWith('application/json')) {
          try {
            js = JSON.parse(buf.toString('utf8'));
          } catch {}
        }
        resolve({
          status: res.statusCode || 0,
          headers: res.headers,
          json: js,
          body: buf,
        });
      });
    });
    req.on('error', reject);
    if (body) {
      req.write(
        Buffer.from(typeof body === 'string' ? body : JSON.stringify(body))
      );
    }
    req.end();
  });
}

async function downloadToFile(url, headers, outPath, _redir = 0) {
  const u = new URL(url);
  const opts = {
    method: 'GET',
    hostname: u.hostname,
    path: u.pathname + u.search,
    headers: { 'User-Agent': 'ghapp-client', ...headers },
  };
  return new Promise((resolve, reject) => {
    const req = https.request(opts, res => {
      // Follow redirects (S3 pre-signed URL)
      if (
        [301, 302, 303, 307, 308].includes(res.statusCode || 0) &&
        res.headers.location
      ) {
        if (_redir > 5)
          return resolve({
            ok: false,
            status: res.statusCode,
            body: 'Too many redirects',
          });
        const loc = res.headers.location;
        // For redirected S3 URL, drop auth headers
        downloadToFile(
          loc,
          { 'User-Agent': 'ghapp-client' },
          outPath,
          _redir + 1
        ).then(resolve, reject);
        return;
      }
      if (res.statusCode !== 200) {
        const chunks = [];
        res.on('data', c => chunks.push(c));
        res.on('end', () => {
          const body = Buffer.concat(chunks).toString('utf8');
          resolve({ ok: false, status: res.statusCode, body });
        });
        return;
      }
      const ws = fs.createWriteStream(outPath);
      res.pipe(ws);
      ws.on('finish', () => resolve({ ok: true }));
      ws.on('error', reject);
    });
    req.on('error', reject);
    req.end();
  });
}

async function main() {
  const args = process.argv.slice(2);
  function getArg(name, def) {
    const i = args.indexOf(name);
    return i >= 0 ? args[i + 1] : def;
  }
  const appId = getArg('--app-id');
  const keyFile = getArg('--key-file');
  const owner = getArg('--owner');
  const repo = getArg('--repo');
  const runId = getArg('--run-id');
  const out = getArg('--out', `run-${runId}.zip`);

  if (!appId || !keyFile || !owner || !repo || !runId) {
    console.error(
      'Usage: --app-id <id> --key-file <pem> --owner <o> --repo <r> --run-id <id> [--out file]'
    );
    process.exit(2);
  }
  const pemPath = path.resolve(keyFile);
  if (!fs.existsSync(pemPath)) {
    console.error(`Private key not found: ${pemPath}`);
    process.exit(2);
  }
  const pem = fs.readFileSync(pemPath, 'utf8');
  const jwt = makeJwt(appId, pem);

  const base = 'https://api.github.com';
  const headersJwt = {
    Authorization: `Bearer ${jwt}`,
    Accept: 'application/vnd.github+json',
    'X-GitHub-Api-Version': '2022-11-28',
  };

  // Resolve installation for repo
  const instResp = await requestJson(
    'GET',
    `${base}/repos/${owner}/${repo}/installation`,
    headersJwt
  );
  if (instResp.status !== 200) {
    console.error(
      'Failed to resolve installation:',
      instResp.status,
      instResp.json || instResp.body?.toString?.()
    );
    process.exit(3);
  }
  const instId = instResp.json.id;
  if (!instId) {
    console.error('No installation id in response');
    process.exit(3);
  }

  // Create installation token
  const tokResp = await requestJson(
    'POST',
    `${base}/app/installations/${instId}/access_tokens`,
    headersJwt,
    {}
  );
  if (tokResp.status !== 201) {
    console.error(
      'Failed to create installation token:',
      tokResp.status,
      tokResp.json || tokResp.body?.toString?.()
    );
    process.exit(4);
  }
  const instToken = tokResp.json.token;
  if (!instToken) {
    console.error('No token in access_tokens response');
    process.exit(4);
  }

  const headersInst = {
    Authorization: `Bearer ${instToken}`,
    Accept: 'application/vnd.github+json',
    'X-GitHub-Api-Version': '2022-11-28',
  };
  const url = `${base}/repos/${owner}/${repo}/actions/runs/${runId}/logs`;
  const dl = await downloadToFile(url, headersInst, out);
  if (!dl.ok) {
    console.error('Download failed:', dl.status, dl.body?.slice?.(0, 300));
    process.exit(5);
  }
  console.log(`Downloaded: ${out}`);
}

main().catch(err => {
  console.error('Unexpected error:', err?.message || err);
  process.exit(1);
});
